src/Runner.Worker/Handlers/CompositeActionHandler.cs

```diff
@@ -26,65 +26,60 @@ public sealed class CompositeActionHandler : Handler, ICompositeActionHandler
 
         public async Task RunAsync(ActionRunStage stage)
         {
-            // Validate args.
+            // Validate args
             Trace.Entering();
             ArgUtil.NotNull(ExecutionContext, nameof(ExecutionContext));
             ArgUtil.NotNull(Inputs, nameof(Inputs));
             ArgUtil.NotNull(Data.Steps, nameof(Data.Steps));
 
-            // Resolve action steps
-            var actionSteps = Data.Steps;
-
-            // Create Context Data to reuse for each composite action step
-            var inputsData = new DictionaryContextData();
-            foreach (var i in Inputs)
-            {
-                inputsData[i.Key] = new StringContextData(i.Value);
-            }
-
-            // Initialize Composite Steps List of Steps
-            var compositeSteps = new List<IStep>();
-
-            // Temporary hack until after M271-ish. After M271-ish the server will never send an empty
-            // context name. Generated context names start with "__"
-            var childScopeName = ExecutionContext.GetFullyQualifiedContextName();
-            if (string.IsNullOrEmpty(childScopeName))
-            {
-                childScopeName = $"__{Guid.NewGuid()}";
-            }
-
-            foreach (Pipelines.ActionStep actionStep in actionSteps)
+            try
             {
-                var actionRunner = HostContext.CreateService<IActionRunner>();
-                actionRunner.Action = actionStep;
-                actionRunner.Stage = stage;
-                actionRunner.Condition = actionStep.Condition;
-
-                var step = ExecutionContext.CreateCompositeStep(childScopeName, actionRunner, inputsData, Environment);
-
-                // Shallow copy github context
-                var gitHubContext = step.ExecutionContext.ExpressionValues["github"] as GitHubContext;
-                ArgUtil.NotNull(gitHubContext, nameof(gitHubContext));
-                gitHubContext = gitHubContext.ShallowCopy();
-                step.ExecutionContext.ExpressionValues["github"] = gitHubContext;
+                // Inputs of the composite step
+                var inputsData = new DictionaryContextData();
+                foreach (var i in Inputs)
+                {
+                    inputsData[i.Key] = new StringContextData(i.Value);
+                }
 
-                // Set GITHUB_ACTION_PATH
-                step.ExecutionContext.SetGitHubContext("action_path", ActionDirectory);
+                // Temporary hack until after M271-ish. After M271-ish the server will never send an empty
+                // context name. Generated context names start with "__"
+                var childScopeName = ExecutionContext.GetFullyQualifiedContextName();
+                if (string.IsNullOrEmpty(childScopeName))
+                {
+                    childScopeName = $"__{Guid.NewGuid()}";
+                }
 
-                compositeSteps.Add(step);
-            }
+                // Create nested steps
```

Comment:
per @TingluoHuang feedback, call this "embedded steps" which lines up w/ the execution context change